upa/src/sdk/events.ts

```diff
@@ -321,3 +323,90 @@ export class VKRegisteredEventGetter extends EventGetterBase<
     };
   }
 }
+
+// Challenge event data is empty
+// We only use it to retrieve associated tx calldata
+export type ChallengeEventOutput = object;
+
+export class ChallengeEventGetter extends EventGetterBase<
+  ChallengeEvent.Event,
+  ChallengeEventOutput
+> {
+  constructor(upa: IUpaVerifier) {
+    super(upa, upa.filters.Challenge());
+  }
+
+  // eslint-disable-next-line
+  parseEvent(ev: TypedEventLog<ChallengeEvent.Event>): ChallengeEventOutput {
+    return {};
+  }
+}
+
+// Event corresponding to successful challenge of a full submission
+// Note: Does not correspond to success of a single call to challenge
+// but rather the successful challenge of a full submission
+// Since a successful challenge happens during a challenge call, the event
+// data is empty
+export class SubmissionChallengeSuccessEventGetter extends EventGetterBase<
+  SubmissionChallengeSuccessEvent.Event,
+  ChallengeEventOutput
+> {
+  constructor(upa: IUpaVerifier) {
+    super(upa, upa.filters.SubmissionChallengeSuccess());
+  }
+
+  // eslint-disable-next-line
+  parseEvent(ev: TypedEventLog<ChallengeEvent.Event>): ChallengeEventOutput {
+    return {};
+  }
+}
+
+export function getCallDataForChallengeTx(
+  verifier: UpaVerifier,
+  tx: ethers.TransactionResponse
+): {
+  circuitId: string;
+  proof: Groth16CompressedProofStruct;
+  publicInputs: bigint[];
+  submissionId: string;
+  dupSubmissionIdx: bigint;
+  proofIdMerkleProof: string[];
+  proofDataMerkleProof: string[];
+} {
+  const challengeFragment = verifier.getFunction("challenge")!.fragment;
+  const decoded = verifier.interface.decodeFunctionData(
+    challengeFragment,
+    tx.data
+  );
+
+  // The decoded data is dynamically indexed.  Extract everything to make it a
+  // concrete struct.
+
+  const circuitId: string = readBytes32(decoded.circuitId);
+  const proof: Groth16CompressedProofStruct = decoded.proof;
+  const publicInputs: bigint[] = decoded.publicInputs.map(
+    (pi: ethers.BigNumberish) => BigInt(pi)
+  );
```

Comment:
Might be able to just do:
```suggestion
  const publicInputs = decoded.publicInputs.map(BigInt);
```

Reply:
Done.